ElrondNetwork/elrond-go
# Implement queries for historical balances (state)

## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
- Node API does not support historical balances lookup yet. That is, it does not support queries such as `address/erd1alice?blockNonce=12345`.
- Rosetta API would greatly benefit from historical balances (e.g. improved robustness, improved debugging)
  
## Proposed Changes
- At the API controllers level (i.e., in `addressGroup`), allow one to provide extra URL parameters (query parameters) for account-related endpoints. These parameters are: `blockNonce`, `blockHash`, `blockRootHash`. They are mutually exclusive.
- At a lower level, in the `node` object, handle the new query options and appropriately delegate the queries to the `accountsRepository`. The `accountsRepository` relies on a new component (see below) to resolve historical queries.
- In addition to `accountsDBApi`, add a new implementation of `AccountsAdapterAPI`: `accountsDBApiWithHistory`. This implementation is able to provide historical state (balance etc.) for the queried account, by recreating the trie at a specified `rootHash`.
- The re-creation of the trie at an older `rootHash` is possible thanks to: https://github.com/ElrondNetwork/elrond-go/pull/4365.
- Historical balances lookup require the following node configuration: `AccountsStatePruningEnabled = false`. The implementation is also subject to (affected by): `NumActivePersisters = K`.

## Testing procedure
- Run Rosetta's `check:data` validation flow, with `historical balances` enabled, on local testnet, then on devnet.
- On Node API, perform a large number of account queries such as: `/address/erd1alice?blockNonce=...`, `/address/erd1alice?blockHash=...`, `/address/erd1alice?blockRootHash=...`. On the response, also check the `blockInfo` field. It should match the provided coordinates.
- For the queries above, also provide bad parameters (e.g. missing blocks).
- For `sender = Alice`, broadcast a transaction that transfers some value. Say this transaction is added in block N. Then, query `address/erd1alice?blockNonce=(N-1)`, `address/erd1alice?blockNonce=(N)` and `address/erd1alice?blockNonce=(N+1)`. Check output.
- On internal testnets, make sure to set `AccountsStatePruningEnabled = false` and `NumActivePersisters = K`, where K should be at least 3 (3 is fine).
